lib/internal/worker/js_transferable.js

```diff
@@ -0,0 +1,31 @@
+'use strict';
+const {
+  messaging_deserialize_symbol,
+  messaging_transfer_symbol,
+  messaging_clone_symbol,
+  messaging_transfer_list_symbol
+} = internalBinding('symbols');
+const {
+  JSTransferable,
+  setDeserializerCreateObjectFunction
+} = internalBinding('messaging');
+
+function setup() {
+  // Register the handler that will be used when deserializing JS-based objects
+  // from .postMessage() calls. The format of `deserializeInfo` is generally
+  // 'module:Constructor', e.g. 'internal/fs/promises:FileHandle'.
+  setDeserializerCreateObjectFunction((deserializeInfo) => {
+    const [ module, ctor ] = deserializeInfo.split(':');
+    const Ctor = require(module)[ctor];
```

Comment:
Should the module name be guarded with a whitelist here? 

I don't think it should since worker threads don't provide a sandbox but I just want to make sure that this isn't exposing a new attack surface somehow.

Reply:
@benjamingr Hm yeah … this isn’t extending capabilities of the calling code in any way, but I agree that it might be good to be careful here. I’ll think about this a bit more.

Reply:
@benjamingr I’ve updated this PR so that this function now verifies that the target constructor is a subclass of `JSTransferable`, and added tests that verify that fact + the fact that this will only load internal modules (which was already the case before).

I think that’s as “locked down” as this can get. If user code makes other internal classes inherit from `JSTransferable` on the receiving side, *and* override the `kTransfer` method on the sending side – which is totally doable in most situation – then that means that you can already make arbitrary changes to Node.js internals and won’t even need to do this.

The one thing I could see happen would be side effects from `require()`ing the internal module itself, which the sending code can still trigger on the receiving side (and which would still typically result in an uncaught exception).

What do you think?

Reply:
> The one thing I could see happen would be side effects from require()ing the internal module itself, which the sending code can still trigger on the receiving side (and which would still typically result in an uncaught exception).

Fwiw that's the part that initially bugged me, but like we both said it doesn't _increase_ the attack surface at all from what I can tell.